pkg/rbac/parser.go

```diff
@@ -253,6 +192,90 @@ func GenerateRoles(ctx *genall.GenerationContext, roleName string) ([]interface{
 	return objs, nil
 }
 
+func GroupRulesByNamespace(ctx *genall.GenerationContext) map[string][]*Rule {
+	rulesByNS := make(map[string][]*Rule)
+	for _, root := range ctx.Roots {
+		markerSet, err := markers.PackageMarkers(ctx.Collector, root)
+		if err != nil {
+			root.AddError(err)
+		}
+
+		// group RBAC markers by namespace
+		for _, markerValue := range markerSet[RuleDefinition.Name] {
+			rule := markerValue.(Rule)
+			namespace := rule.Namespace
+			rulesByNS[namespace] = append(rulesByNS[namespace], &rule)
+		}
+	}
+
+	return rulesByNS
+}
+
+// insertRule inserts a rule into a destination slice, deduplicating via the
+// Subsumes function or merging verbs if appropriate
+func insertRule(dest []*NormalizedRule, it *NormalizedRule) []*NormalizedRule {
+	// this is not going to be very fast but the set of rules should always be small
+	mergeWith := -1
+	for ix, other := range dest {
+		if other.Subsumes(it) {
+			// not needed; another rule handles this case
+			return dest
+		}
+
+		if it.Subsumes(other) {
+			// rebuild whole list
+			result := []*NormalizedRule{it}
+			for _, d := range dest {
+				result = insertRule(result, d)
+			}
```

Comment:
It could, but since `it.Subsumes(other)` and we start the new list with `it` being contained, then `other` will be omitted by the previous check above. Explicitly skipping it makes the code a bit messier.

Reply:
What you are doing here is both harder to read and less eficient than:
```
func remove(s []int, i int) []int {
    s[i] = s[len(s)-1]
    return s[:len(s)-1]
}
return append(remove(dest, idx), it)
```

Reply:
@alvaroaleman that doesn’t result in the quite same outcome; if Rule X is superset of Rule Y in the list, it could also be a superset of Rule Z (and others), so we still need to go through and check the remainder of the list.